File: H-teams/sol.py
# ai translated from cpp

import sys
from collections import defaultdict
 
def solve():
    n, m, ka, kb = map(int, sys.stdin.readline().split())
    skill = list(map(int, sys.stdin.readline().split()))
 
    total_skill = sum(skill)
 
    adj = defaultdict(list)
    for _ in range(m):
        u, v = map(int, sys.stdin.readline().split())
        u -= 1
        v -= 1
        adj[u].append(v)
        adj[v].append(u)
 
    fgs = []
    vis = [False] * n
 
    def dfs(x):
        if vis[x]:
            return
        vis[x] = True
 
        fgs[-1]['players'].append(x)
        fgs[-1]['tot_skill'] += skill[x]
 
        for ch in adj[x]:
            dfs(ch)
 
    for x in range(n):
        if not vis[x]:
            fgs.append({'players': [], 'tot_skill': 0})
            dfs(x)
 
    DNE = (-10, -10, -10)
    ORIGIN = (-1, -1, -1)
 
    dp = [[[DNE for _ in range(ka + 1)] for _ in range(n // 2 + 1)] for _ in range(len(fgs) + 1)]
    dp[0][0][0] = ORIGIN
 
    for i in range(1, len(fgs) + 1):
        for v in range(n // 2 + 1):
            for w in range(ka + 1):
                dp[i][v][w] = dp[i - 1][v][w]
 
                nv = v - len(fgs[i - 1]['players'])
                nw = w - fgs[i - 1]['tot_skill']
                if nv >= 0 and nw >= 0 and dp[i][v][w] == DNE and dp[i - 1][nv][nw] != DNE:
                    dp[i][v][w] = (i - 1, nv, nw)
 
    for i in range(len(fgs) + 1):
        for wa in range(ka + 1):
            if dp[i][n // 2][wa] != DNE:
                wb = total_skill - wa
 
                if wb <= kb:
                    team_a_friendgroups = []
 
                    s = dp[i][n // 2][wa]
                    while s != ORIGIN:
                        team_a_friendgroups.append(s[0])
                        s = dp[s[0]][s[1]][s[2]]
 
                    player_team_a = [False] * n
                    for fgix in team_a_friendgroups:
                        for player in fgs[fgix]['players']:
                            player_team_a[player] = True
 
                    team_a_players = []
                    for j in range(n):
                        if player_team_a[j]:
                            team_a_players.append(j)
 
                    for j in range(n // 2):
                        print(team_a_players[j] + 1, end=" \n"[j == n // 2 - 1])
 
                    return
 
    print("-1")
 
if __name__ == "__main__":
    solve()

File: H-teams/sol.cpp
#include<bits/stdc++.h>
#pragma GCC optimize("O3,unroll-loops")
using namespace std;

#define int long long int

struct friendship_group {
        vector<int> players;
        int tot_skill;
};

struct state {
        int i,v,w;
        bool operator==(const state &other) const {
                return i==other.i && v==other.v && w==other.w;
        }
        bool operator!=(const state &other) const {
                return !(*this == other);
        }
};

void solve() {
        /**
         * BLOCK: INPUT
         * IN: standard input
         * OUT: player skills
         */
        int n,m,ka,kb;cin>>n>>m>>ka>>kb;
        vector<int> skill(n); for (int i=0;i<n;i++) cin >> skill[i];

        /**
         * BLOCK: SKILL SUMMER
         * IN: player skills
         * OUT: sum of all player skills
         */
        int total_skill = 0; for (int i=0;i<n;i++) total_skill += skill[i];

        /**
         * BLOCK: ADJ. LIST CONSTRUCTOR
         * IN: standard input
         * OUT: adjacency list
         */
        vector<vector<int>> adj(n);
        for (int i=0;i<m;i++) {
                int u,v;cin>>u>>v;u--;v--;  // 0-index
                adj[u].push_back(v);
                adj[v].push_back(u);
        }

        /**
         * BLOCK: CC FINDER
         * IN: adj. list, player skills
         * OUT: friendship groups
         */
        vector<friendship_group> fgs;
        vector<bool> vis(n);
        function<void(int)> dfs = [&](int x) {
                if (vis[x]) return;
                vis[x] = true;

                fgs.rbegin()->players.push_back(x);
                fgs.rbegin()->tot_skill += skill[x];

                for (int ch : adj[x]) dfs(ch);
        };
        for (int x=0;x<n;x++) {
                if (!vis[x]) {
                        fgs.push_back({{}, 0});
                        dfs(x);
                }
        }

        /**
         * BLOCK: KNAPSACK DP
         * IN: friendship groups
         * OUT: A knapsack dp array that:
         *      1. shows whether the first i friendship groups can produce
         *              exactly v players and w skill
         *      2. if so, gives the last state where a friendship group was
         *              added to team A to reach the current state
         */
        state DNE = {-10, -10, -10};  // arbitrary negative value
        state ORIGIN = {-1, -1, -1};

        vector<vector<vector<state>>> dp(fgs.size()+1, vector<vector<state>>(n/2+1, vector<state>(ka+1, DNE)));

        dp[0][0][0] = ORIGIN;
        for (int i=1;i<=fgs.size();i++) {
                for (int v=0;v<=n/2;v++) {  // v --> players
                        for (int w=0;w<=ka;w++) {  // w --> skill
                                dp[i][v][w] = dp[i-1][v][w];

                                int nv = v-fgs[i-1].players.size(), nw = w-fgs[i-1].tot_skill;
                                if (
                                        nv >= 0 && nw >= 0 &&
                                        dp[i][v][w] == DNE &&
                                        dp[i-1][nv][nw] != DNE
                                ) dp[i][v][w] = {i-1, nv, nw};
                        }
                }
        }

        /**
         * BLOCK: SOLUTION UNPACKER
         * IN: knapsack dp array
         * OUT: team a players (printed)
         * METHOD: check if team A configs work; track backwards from working solution
         */
        for (int i=1;i<=fgs.size();i++) {
                for (int wa=ka;wa>=1;wa--) {
                        if (dp[i][n/2][wa] != DNE) {
                                int wb = total_skill - wa;

                                if (wb <= kb) {
                                        vector<int> team_a_friendgroups;
                                        
                                        state s = dp[i][n/2][wa];
                                        while (s != ORIGIN) {
                                                team_a_friendgroups.push_back(s.i);
                                                s = dp[s.i][s.v][s.w];
                                        }

                                        vector<bool> player_team_a(n);
                                        for (int fgix : team_a_friendgroups) {
                                                for (int player : fgs[fgix].players) {
                                                        player_team_a[player] = true;
                                                }
                                        }

                                        vector<int> team_a_players, team_b_players;
                                        for (int j=0;j<n;j++) {
                                                if (player_team_a[j]) team_a_players.push_back(j);
                                        }
                                        
                                        for (int j=0;j<n/2;j++)
                                                cout << (team_a_players[j] + 1) << " \n"[j==n/2-1];
                                        
                                        return;
                                }
                        }
                }
        }

        cout << "-1\n";
}

int32_t main() {
        ios_base::sync_with_stdio(false);
        cin.tie(0);
        solve();
}
